### PR TITLE
Reassociation pass; two more strength-reduction optimizations

### LINKED ISSUE #2585: ConstantFolding not folding as expected
For something like this:
```p4
h.data = h.data2 - 8 - 8 - 2 - 16;
```
I would expect `ConstantFolding` to fold this to:
```p4
h.data = h.data2 - 34;
```
however, it does not fold the constants at all.

If parens are added:
```p4
h.data = h.data2 - (8 + 8 + 2 + 16);
```
then the constants are folded as expected.

test: 
[constant-fold.p4.txt](https://github.com/p4lang/p4c/files/5375005/constant-fold.p4.txt)

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Fixes #2585, #1149

### CODE REVIEW

**Comment by jafingerhut** [comment]
All of the updated expected outputs look reasonable to me, but as usual, I'm not a good one for reviewing p4c pass changes.

A question: I assume that saturating add and subtract operations are not treated as associative by p4c?

I know that `(a - constant1)` expressions are often (maybe always?) transformed by p4c into a functionally equivalent expression `(a + constant2)` where `constant2` is, I believe, an unsigned representation of the 2's complement of `constant1`.  I suppose that such a transformation is never made for expressions like `(a |-| constant1)` ?

**Comment by mbudiu-vmw** [comment]
I believe that saturated operations are not associative. a + 1 + (-1) is not the same as a if a is max.
Yes, the saturated operations have different opcodes, and I don't think we have many optimizations (if any) for them.
We could probably do some strength-reduction, though.

**Comment by mbudiu-vmw** [comment]
Actually I think this is really one of the simplest passes one could write, so perhaps you would like to take a look at the implementation too.

**Review by ChrisDodd — APPROVED** [review]

File: frontends/p4/reassociation.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    { return reassociate(expr); }
+    const IR::Node* postorder(IR::Mul* expr) override
+    { return reassociate(expr); }
+    const IR::Node* postorder(IR::BOr* expr) override
+    { return reassociate(expr); }
+    const IR::Node* postorder(IR::BAnd* expr) override
+    { return reassociate(expr); }
+    const IR::Node* postorder(IR::BXor* expr) override
+    { return reassociate(expr); }
+};
```
Comment (jphurl) [inline frontends/p4/reassociation.h:44]: How does the testcase work? since we don't visit `IR::Sub`

Comment (jfingerh) [reply frontends/p4/reassociation.h:44]: If I understand your question correctly, and what the compiler currently does, it transforms all expressions of the form `(variable - constant1)` into another expression guaranteed to evaluate to the same result that looks like `(variable + constant2)`, where I believe `constant2` is the 2's complement negation of `constant1`.  After that transformation, there are no more subtractions by constant any more.

Comment (jphurl) [reply frontends/p4/reassociation.h:44]: ah ok thank you!

**Pull request merged**